src/libraries/Microsoft.PowerFx.Interpreter/UDFHelper.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Interpreter
+{
+    internal class UDFHelper
+    {
+        public class LazyCheck
+        {
+            internal string _expressionText;
+            internal RecordType _parameterType;
+            internal ParserOptions _options;
+            internal Engine _engine;
+
+            public LazyCheck(Engine engine, string expressionText, RecordType parameterType = null, ParserOptions options = null)
+            {
+                _engine = engine;
+                _expressionText = expressionText;
+                _parameterType = parameterType;
+                _options = options;
+            }
+
+            public CheckResult Get() => _engine.Check(_expressionText, _parameterType, _options);
+        }
+
+        // would also be a good generic helper 
+        private static NamedValue[] Zip(NamedFormulaType[] parameters, FormulaValue[] args)
+        {
+            if (parameters.Length != args.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            var result = new NamedValue[args.Length];
+            for (var i = 0; i < args.Length; i++)
+            {
+                result[i] = new NamedValue(parameters[i].Name, args[i]);
+            }
+
+            return result;
+        }
+        
+        public class UDFDefinition
+        {
+            internal string _name;
+            internal string _body;
+            internal FormulaType _returnType;
+            internal NamedFormulaType[] _parameters;
+
+            public UDFDefinition(string name, string body, FormulaType returnType, params NamedFormulaType[] parameters)
+            {
+                _name = name;
+                _body = body;
+                _returnType = returnType;
+                _parameters = parameters;
+            }
+        }
+
+        internal class UDFLazyBinder
+        {
+            internal UDFTexlFunction _function;
+            internal string _name;
+            internal List<ExpressionError> _expressionError = new List<ExpressionError>();
+
+            public UDFLazyBinder(UDFTexlFunction texlFunction, string name)
+            { 
+                _function = texlFunction;
+                _name = name;
+            }
+
+            // Used when trying to define a function, there is an error before binding, such as the function name already being defined.
+            public UDFLazyBinder(ExpressionError expressionError, string name)
+            {
+                _expressionError.Add(expressionError);
+                _name = name;
+            }
+
+            public string GetName()
```

Comment:
>GetName [](http://example.com/codeflow?start=26&length=7)

the general C# preference here is for a property

`public string Name {get;}` 

over GetName+ _name

Reply:
Fixed